File: BgMeetup/src/main/java/com/bgmeetup/backend/service/EventService.java
package com.bgmeetup.backend.service;

import com.bgmeetup.backend.domain.*;
import com.bgmeetup.backend.dto.*;
import com.bgmeetup.backend.enums.InviteStatus;
import com.bgmeetup.backend.exceptions.EntityNotFoundException;
import com.bgmeetup.backend.mapper.EventMapper;
import com.bgmeetup.backend.mapper.EventParticipantMapper;
import com.bgmeetup.backend.mapper.LeaderboardScoreMapper;
import com.bgmeetup.backend.repository.EventRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class EventService {

    private final EventMapper eventMapper;
    private final EventRepository eventRepository;
    private final EventParticipantMapper eventParticipantMapper;
    private final UserService userService;
    private final GameService gameService;
    private final LeaderboardScoreMapper leaderboardScoreMapper;

    @Autowired
    public EventService(EventMapper eventMapper,
                        EventRepository eventRepository,
                        EventParticipantMapper eventParticipantMapper,
                        UserService userService,
                        GameService gameService,
                        LeaderboardScoreMapper leaderboardScoreMapper) {
        this.eventMapper = eventMapper;
        this.eventRepository = eventRepository;
        this.eventParticipantMapper = eventParticipantMapper;
        this.userService = userService;
        this.gameService = gameService;
        this.leaderboardScoreMapper = leaderboardScoreMapper;
    }

    public EventDto get(String eventId) {
        var event = eventRepository.get(eventId).orElseThrow(()-> new EntityNotFoundException("Event"));
        String dateString = event.getDate().toString();
        var dateParts  = dateString.split("T");
        event.setDateString(dateParts[0] + " " + dateParts[1]);
        event.setDate(null);
        var host = userService.get(event.getHostId().toString());
        event.setHostName(host.getLastName() + " " + host.getFirstName());
        var participants = eventRepository.getParticipants(event.getId().toString());
        var participantsCount = participants.size();
        event.setParticipantsCount(participantsCount);
        return event;
    }

    public List<EventDto> getAll(String userId) {
        var events = eventRepository.getAll();
        for (EventDto event : events) {
            String dateString = event.getDate().toString();
            var dateParts  = dateString.split("T");
            event.setDateString(dateParts[0] + " " + dateParts[1]);
            event.setDate(null);
            var host = userService.get(event.getHostId().toString());
            event.setHostName(host.getLastName() + " " + host.getFirstName());
            var participants = eventRepository.getParticipants(event.getId().toString());
            var participantsCount = participants.stream().filter(p -> p.getStatus() == InviteStatus.Accepted.getValue()).count();
            event.setParticipantsCount((int)participantsCount);
            var participant = participants.stream().filter(p -> p.getParticipantId().toString().equals(userId)).findFirst();
            if(participant.isPresent()){
                event.setStatus(participant.get().getStatus());
                var inviter = userService.get(participant.get().getInviterId().toString());
                event.setInvitedBy(inviter.getLastName() + " " + inviter.getFirstName());
            }
            else
                event.setStatus(InviteStatus.NotInvited.getValue());
        }

        return events;
    }

    @Transactional(rollbackFor = Exception.class)
    public SaveResult create(EventDto request) {
        Event event = eventMapper.toEntity(request);
        var eventId = eventRepository.create(event);
        var host = userService.get(request.getHostId().toString());
        var participant = new EventParticipant(eventId, event.getHostId(), event.getHostId(), host.getEmail(), InviteStatus.Accepted.getValue(), false);
        return eventRepository.invite(participant);
    }

    @Transactional(rollbackFor = Exception.class)
    public SaveResult update(EventDto request) {
        Event event = eventMapper.toEntity(request);
        return eventRepository.update(event);
    }

    public SaveResult join(String eventId, String userId) {
        return eventRepository.join(eventId, userId);
    }

    public SaveResult leave(String eventId, String userId) {
        return eventRepository.leave(eventId, userId);
    }

    public SaveResult cancel(String eventId) {
        return eventRepository.cancel(eventId);
    }

    public SaveResult confirm(String eventId) {
        return eventRepository.confirm(eventId);
    }

    @Transactional(rollbackFor = Exception.class)
    public SaveResult invite(EventParticipantDto request) {
        var dbParticipant = eventRepository.getParticipant(request.getEventId().toString(), request.getEmail());

        if(request.getParticipantId() == null){
            var user = userService.getByEmail(request.getEmail());
            request.setParticipantId(user.getId());
        }
        else {
            var user = userService.get(request.getParticipantId().toString());
            request.setEmail(user.getEmail());
        }
        EventParticipant participant = eventParticipantMapper.toEntity(request);
        if (dbParticipant.isPresent())
            return eventRepository.acceptInvitation(request.getEventId().toString(), request.getParticipantId().toString());
        else
            return eventRepository.invite(participant);
    }

    public SaveResult acceptInvitation(String eventId, String userId) {
        return eventRepository.acceptInvitation(eventId, userId);
    }

    public SaveResult declineInvitation(String eventId, String userId) {
        return eventRepository.declineInvitation(eventId, userId);
    }

    public List<EventParticipantDto> getParticipants(String eventId) {
        var participants = eventRepository.getParticipants(eventId);
        for (EventParticipantDto participant : participants) {
            var user = userService.get(participant.getParticipantId().toString());
            if(participant.getParticipantName() == null)
                participant.setParticipantName(user.getEmail());
            else
                participant.setParticipantName(user.getLastName() + " " + user.getFirstName());
        }

        return participants;
    }

    public SaveResult checkIn(String eventId, String userId) {
        return eventRepository.checkIn(eventId, userId);
    }

    @Transactional(rollbackFor = Exception.class)
    public SaveResult submitLeaderboard(List<LeaderboardScoreDto> requests) {
        List<LeaderboardScore> games = leaderboardScoreMapper.toEntityList(requests);
        return eventRepository.submitLeaderboard(games);
    }

    public List<LeaderboardScoreDto> getLeaderboard(String eventId) {
        var leaderboard = eventRepository.getLeaderboard(eventId);
        for (LeaderboardScoreDto score : leaderboard) {
            var participant = userService.get(score.getParticipantId().toString());
            score.setParticipantName(participant.getLastName() + " " + participant.getFirstName());
            var game = gameService.get(score.getGameId().toString());
            score.setTitle(game.getTitle());
        }

        return leaderboard;
    }

    @Transactional(rollbackFor = Exception.class)
    public SaveResult submitFeedback(List<Feedback> requests) {
        return eventRepository.submitFeedback(requests);
    }

    public List<Feedback> getFeedback(String eventId) {
        return eventRepository.getFeedback(eventId);
    }
}

File: BgMeetup-Frontend/src/app/services/event.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { map } from "rxjs/operators";
import { environment } from "../../environments/environment";
import { EventModel } from "../models/event.model";
import { EventParticipantModel } from "../models/eventParticipant.model";
import { LeaderboardScoreModel } from "../models/leaderboardScore.model";
import { ProposedGameModel } from "../models/proposedGame.model";
import { SaveResult } from "../models/saveResult";
import { BGGService } from "./bgg.service";

@Injectable({ providedIn: 'root' })
export class EventService {
  req = new XMLHttpRequest();

  constructor(
    private http: HttpClient,
    private bggService: BGGService) { }

  getEvent(id: any): Observable<EventModel> {
    return this.http.get<EventModel>(environment.apiUrl + '/events/get/' + id)
      .pipe(map(returnValue => {

        return returnValue;
      }));
  }

  getEvents(id: any): Observable<EventModel[]> {
    return this.http.get<EventModel[]>(environment.apiUrl + '/events/all/' + id)
      .pipe(map(returnValue => {

        return returnValue;
      }));
  }

  create(eventModel: EventModel) {
    return this.http.post<any>(environment.apiUrl + '/events/create', eventModel)
      .pipe(map(returnValue => {
        return returnValue;
      }));
  }

  update(eventModel: EventModel) {
    return this.http.post<any>(environment.apiUrl + '/events/update', eventModel)
      .pipe(returnValue => {
        return returnValue;
      });
  }

  join(eventId: any, userId: any): Observable<SaveResult> {
    return this.http.get<SaveResult>(environment.apiUrl + '/events/join/eventId=' + eventId + "/userId=" + userId)
      .pipe(map(returnValue => {

        return returnValue;
      }));
  }

  leave(eventId: any, userId: any): Observable<SaveResult> {
    return this.http.get<SaveResult>(environment.apiUrl + '/events/leave/eventId=' + eventId + "/userId=" + userId)
      .pipe(map(returnValue => {

        return returnValue;
      }));
  }

  cancel(eventId: any): Observable<SaveResult> {
    return this.http.get<SaveResult>(environment.apiUrl + '/events/cancel/' + eventId)
      .pipe(map(returnValue => {

        return returnValue;
      }));
  }

  getParticipants(id: any): Observable<EventParticipantModel[]> {
    return this.http.get<EventParticipantModel[]>(environment.apiUrl + '/events/getParticipants/' + id);
  }

  invite(participantModel: EventParticipantModel) {
    return this.http.post<any>(environment.apiUrl + '/events/invite', participantModel)
      .pipe(returnValue => {
        return returnValue;
      });
  }

  acceptInvitation(eventId: any, userId: any): Observable<any> {
    return this.http.get<any>(environment.apiUrl + '/events/acceptInvitation/eventId=' + eventId + "/userId=" + userId);
  }

  declineInvitation(eventId: any, userId: any): Observable<any> {
    return this.http.get<any>(environment.apiUrl + '/events/declineInvitation/eventId=' + eventId + "/userId=" + userId);
  }

  chooseGames(games: any[], eventId: any) {
    var saveResult = new SaveResult();
    saveResult.result = true;
    saveResult.errors = null;

    return saveResult;
  }

  checkIn(eventId: any, userId: any) {
    return this.http.get<any>(environment.apiUrl + '/events/checkIn/eventId=' + eventId + "/userId=" + userId);
  }

  confirm(eventId: any): Observable<SaveResult> {
    return this.http.get<SaveResult>(environment.apiUrl + '/events/confirm/' + eventId)
      .pipe(map(returnValue => {

        return returnValue;
      }));
  }

  submitLeaderboard(leaderboard: any[]) {
    return this.http.post<any>(environment.apiUrl + '/events/submitLeaderboard', leaderboard)
      .pipe(map(returnValue => {
        return returnValue;
      }));
  }

  getLeaderboard(eventId: any): Observable<LeaderboardScoreModel[]> {
    return this.http.get<LeaderboardScoreModel[]>(environment.apiUrl + '/events/getLeaderboard/' + eventId);
  }
}
